### ISSUE TITLE
[QST] Continued problems with undefined symbols on import

### ISSUE THREAD

**Issue by VisionaryMind** [rapidsai/cudf#7460]
I have seen several closed threads on this question, but after compiling cudf 0.18.0 and building the python libraries, there still seems to be a problem with unresolved symbols, under specific circumstances. The error is not specific to a particular library (all are affected, as I explain in further detail below):

>>> import cudf
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/__init__.py", line 11, in <module>
    from cudf import core, datasets, testing
  File "/usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/core/__init__.py", line 3, in <module>
    from cudf.core import buffer, column, column_accessor, common
  File "/usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/core/column/__init__.py", line 3, in <module>
    from cudf.core.column.categorical import CategoricalColumn
  File "/usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/core/column/categorical.py", line 22, in <module>
    from cudf import _lib as libcudf
  File "/usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/_lib/__init__.py", line 4, in <module>
    from . import (
ImportError: /usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/_lib/avro.cpython-39-aarch64-linux-gnu.so: undefined symbol: _ZN4cudf2io9read_avroERKNS0_19avro_reader_optionsEPN3rmm2mr22device_memory_resourceE

<details>
<summary>Current Environment</summary>

LD_LIBRARY_PATH=/home/visionarymind/melodic_ws/devel/lib:/usr/local/cuda-10.2/lib64:/usr/local/lib:
PYTHONPATH=/home/visionarymind/melodic_ws/devel/lib/python3.6/site-packages
PATH=/home/visionarymind/melodic_ws/devel/bin:/usr/local/cuda-10.2/bin:/home/visionarymind/bin:/home/visionarymind/.local/bin:/home/visionarymind/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
PKG_CONFIG_PATH=/home/visionarymind/melodic_ws/devel/lib/pkgconfig
</details>

According to Issue [#7062](https://github.com/rapidsai/cudf/issues/7062), I changed L114 from ["cudf"] to ["cudf", "cudf_base", "cudf_ast", "cudf_comms", "cudf_hash", "cudf_interop", "cudf_io", "cudf_join", "cudf_merge", "cudf_partitioning", "cudf_reductions", "cudf_replace", "cudf_reshape", "cudf_rolling", "cudf_transpose"].

This has no effect on the problem. Further, if I ldd the above mentioned library, it's clearly linked to libcudart:

	linux-vdso.so.1 (0x0000007f8f505000)
	libgtk3-nocsd.so.0 (0x0000007f8f441000)
	libcudart.so.10.2 => /usr/local/cuda-10.2/lib64/libcudart.so.10.2 (0x0000007f8f3cd000)
	libstdc++.so.6 (0x0000007f8f239000)
	libgcc_s.so.1 => /lib/aarch64-linux-gnu/libgcc_s.so.1 (0x0000007f8f1c0000)
	libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000007f8f067000)
	/lib/ld-linux-aarch64.so.1 (0x0000007f8f4d9000)
	libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000007f8f052000)
	libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000007f8f026000)
	librt.so.1 => /lib/aarch64-linux-gnu/librt.so.1 (0x0000007f8f00f000)
	libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000007f8ef56000)

And If I remove items from this list in cudf/_lib/__init__.py:

from . import (
    avro,
    binaryop,
    concat,
    copying,
    csv,
    datetime,
    filling,
    gpuarrow,
...

It will not find any of them. For example, if I remove avro, it will fail on binaryop:

ImportError: /usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/_lib/binaryop.cpython-39-aarch64-linux-gnu.so: undefined symbol: _ZN4cudf16binary_operationERKNS_11column_viewERKNS_6scalarENS_15binary_operatorENS_9data_typeEPN3rmm2mr22device_memory_resourceE

There are zero compile errors and libraries appear to be linked to libcudart properly. Within the /usr/local/cuda-10.2/lib64/ directory:

libcudart.so
libcudart.so.10.2
libcudart.so.10.2.89
libcudart_static.a

Given that I have compiled and re-compiled this likely over 20 times at this point, could you provide some additional advice on how to allow cudf to find these links properly?

**Comment by kkraus14** [comment]
Hi @VisionaryMind, it looks like the error isn't from linking to `libcudart` in any way, but is getting an undefined symbol from a libcudf binary operator API. Normally this happens from an out of date `rmm` or it finding an incorrect `libcudf` installation.

Could you make sure to pull and build/install the latest `rmm` and then try rebuilding `libcudf` / `cudf` with it?

If you're still getting the above error it's likely that a different `libcudf` binary is being found instead of the one you're building. Could you run `ldd /usr/local/lib/python3.9/site-packages/cudf-0.19.0a0+124.g580f9a200e.dirty-py3.9-linux-aarch64.egg/cudf/_lib/binaryop.cpython-39-aarch64-linux-gnu.so` and share the output here?

**Comment by VisionaryMind** [comment]
I have come to the conclusion that cuDF and other similar CUDA-enabled libraries are best not implemented on Jetson edge devices. We have seen consistent problems with version inconsistencies, overlapping libraries, and other such problems that plague the Jetpack implementation for these devices, increasing compile and testing time from hours or days to months. We've been running benchmarks since last October, and cuDF is the last in a line of experiments that prove Jetson is not up to the task.

We were eventually able to compile and install CuPy during testing, for example, but it runs much slower on Xavier AGX than CPU-based operations on a moderately powerful desktop computer. For this reason, it is our conclusion that Jetson (Nano, TX, Xavier, etc.) GPU-computing capacity is not well suited to big data / analytics or even basic machine-learning pipelines. The question arises how they can be practically implemented, but that is beside the point here. Conda cannot be installed on such devices, making the version chaos much more of an issue.

That being said, we have pulled the Xavier AGX from our ML / AI workflow, as even with CUDA-compiled libraries, it is too slow to be of practical use. Instead, we have installed cuDF and associated libraries using Conda on a workstation with GeForce RTX 2070, and it is loading without a problem. We are running into another issue that looks, at least initially, like cuDF is unable to create dataframes from NumPy arrays or Pandas dataframes generated from NumPy arrays. I will post this issue in another thread.

I am closing this query, as it appears the problem is simply the suboptimally designed Jetson architecture and its lack of platform governance. For others reading this thread, save yourself some time and do not attempt to run machine learning / AI pipelines through edge devices such as Jetson. They are a waste of time, on many levels.
